### ISSUE TITLE
action code duplication reduction feature

### ISSUE THREAD

**Issue by Xeverous** [Xeverous/filter_spirit#3]
The goal: we want a feature that eliminates code duplication of actions that frequently are used together (eg text/border/background color, beam + minimap icon).

We also want to be careful with its complexity:

> **/u/besplash:** I felt like your initial example code was very easy and intuitive to use, but your macros are making it more complex than it has to be. The Tag system from FilterBlast on the other hand doesn't need any new kind of programming knowledge which would be way better for people that have 0 programming experience like most of the playerbase.

___

Example code, containing duplicated actions:

```
currency = RGB(213, 159, 15) # (yellow)

Class "Currency" {
    # some common styles for all currency here...

    BaseType "Chaos Orb" {
        SetTextColor black
        SetBackgroundColor currency
        SetFontSize 42
        Show
    }

    # other blocks between [...]

    BaseType "Orb of Fusing" {
        SetTextColor black
        SetBackgroundColor currency # again with text color, duplicated in chaos orb
        Show
    }

}
```

If there was no "other blocks between [...]", we could just create another scope that inherits these but in this scenario we do not have A, B, B, C order but something like A, B, C, B. Bs are separate so we can not enclose them in `{ [...] }`.
___

action code duplication reduction feature - current idea:

```
currency_colors = {
    SetTextColor black
    SetBackgroundColor currency
}

beam_and_icon = {
    SetBeam yellow
    SetMinimapIcon MinimapIcon(0, star, yellow)
    # also possible here: ApplyTags ...
}

# [...]    
BaseType ["Orb of Fusing", "Orb of Alchemy"] {
    ApplyTags [currency_colors, beam_and_icon]
    SetFontSize 42
    Show
}

BaseType ["Jeweller's Orb"] {
    ApplyTags [currency_colors]
    Show
}
```

I think this syntax is faily simple while being consistent with the rest of the grammar. **Tags could only be used to group actions, no conditions**. Reason:

How filters work: an item goes through it and receives the style of the first block that matches specified conditions, all further blocks are ignored. So you want the most restrictive blocks first (most nested), then decreasing in strictness unitll some generic rule like "hide all leftover magic and rare garbage" hides everything at the bottom. Allowing to override actions is no problem: these do not collide. Filter with styles/colors 1, 2, 3, 4 ..., 97, 98, 99 will work just like 1, 97, 99, 50, 4, 98, 3. But for conditions - if you have blocks in order like A, Z, K, D, Y, C (where A means most restrictive block and Z least restrictive) everything past Z will not work because some items will be catched by A and everything else will be caught by Z, leaving no items for further blocks. If we disallow overriding/tagging conditions in nested blocks, we guuarantee that the filter structure has never "the A, Z, B issue" and always goes from most-specific-item to least-specific-item.

___

Looking for any opinions & propositions:

- How this feature should be named? **Tags** do not seem to be a good name for me, this term should be saved for features related to filter strictness/variants. Maybe **compound actions**, **grouped actions**?
- Grouped actions are applied to blocks by something like `Apply [grp1, grp2]` syntax. Grouped actions can also be applied to other grouped actions. 
- Block grouping conditions, allow only to group actions. See reasoning above. Is this fine?

**Comment by Xeverous** [comment]
Implemented in e6b736c1598061a28cf6e7b08767608f5c765bb1. Needs documentation.

**Comment by Xeverous** [comment]
Doc added in 0473a82df2e1282812fce0c4a32ed9544e433674.
